soroban-env-host/src/fees.rs

```diff
@@ -116,10 +158,83 @@ pub fn compute_transaction_resource_fee(
         .saturating_add(historical_fee)
         .saturating_add(bandwidth_fee);
 
-    (
-        refundable_fee.saturating_add(non_refundable_fee),
-        refundable_fee,
-    )
+    (non_refundable_fee, refundable_fee)
+}
+
+/// Computes the total rent-related fee for the provided ledger entry changes.
+///
+/// The rent-related fees consist of the fees for rent bumps and fees for
+/// increasing the entry size (with or without rent bump).
+///
+/// This cannot handle unsantized inputs and relies on sane configuration and
+/// ledger changes. This is due to the fact that rent is managed automatically
+/// wihtout user-provided inputs.
+pub fn compute_rent_fee(
+    changed_entries: &Vec<LedgerEntryRentChange>,
+    fee_config: &RentFeeConfiguration,
+    current_ledger_seq: u32,
+) -> i64 {
+    let mut fee = 0;
+    for e in changed_entries {
+        fee += rent_fee_per_entry_change(e, fee_config, current_ledger_seq);
+    }
+    fee
+}
+
+fn rent_fee_per_entry_change(
+    entry_change: &LedgerEntryRentChange,
+    fee_config: &RentFeeConfiguration,
+    current_ledger: u32,
+) -> i64 {
+    let mut fee = 0;
+    // Pay for the rent extension (if any).
+    if entry_change.new_expiration_ledger > entry_change.old_expiration_ledger {
+        fee += rent_fee_for_size_and_ledgers(
+            entry_change.is_persistent,
+            // New portion of rent is payed for the new size of the entry.
+            entry_change.new_size_bytes,
+            // Rent should be covered until `old_expiration_ledger` (or start
+            // from the current ledger for new entries), so don't include it
+            // into the number of rent ledgers.
+            entry_change.new_expiration_ledger
+                - entry_change.old_expiration_ledger.max(current_ledger - 1),
+            fee_config,
+        );
+    }
+    // Pay for the entry size increase (if any).
+    if entry_change.new_size_bytes > entry_change.old_size_bytes && entry_change.old_size_bytes > 0
+    {
+        fee += rent_fee_for_size_and_ledgers(
+            entry_change.is_persistent,
+            // Pay only for the size increase.
+            entry_change.new_size_bytes - entry_change.old_size_bytes,
+            // Cover ledger interval [current; old], as (old, new] is already
+            // covered above for the whole new size.
+            entry_change.old_expiration_ledger - current_ledger + 1,
```

Comment:
Won't this be negative if current_ledger > old_expiration_ledger?

Reply:
Isn't that an invariant that we don't bump expired entries? Other part of the logic also rely on that.

As I've mentioned in the docs, this function doesn't try hard to protect against bad inputs, as it's not meant to be called on any user inputs (so bumping an expired entry shouldn't happen).

Reply:
We restore expired entries by bumping them.

Reply:
The restored entries are new entries from the rent perspective, so old size should be `0` as well (as we don't "back-charge" rent).

Reply:
I thought we'd back-charge for entries that were expired (not evicted). After talking to Dima, it sounds like we won't be doing this and will treat expired and evicted the same so that's that we do everywhere else. (CC @SirTyson).